### ISSUE TITLE
Integration with obsidian-icon-folder

### ISSUE THREAD

**Issue by a-rock-lobster** [SebastianMC/obsidian-custom-sort#60]
Hullo,

I have no idea how feasible or challenging this would be but I thought it would useful if it were possible to specify a file or folder icon (perhaps via the icons name) in the sortspec. 

I have been using these icons to help me scan through my list of files. 
https://github.com/FlorianWoelki/obsidian-icon-folder

For example, one might use the bookmark icon to indicate certain files are book chapters or sections and maybe it would be helpful for them to cluster together (then I imagine alpha-numeric would kick in to order the cluster). 

Just a thought. 

Thanks for sharing your work,

**Comment by SebastianMC** [comment]
Hi @a-rock-lobster,

I'm also a user of `obsidian-icon-folder` and thus I know how it works from UI side, user perspective
I've quickly examined internals of that plugin and it looks it exposes the data necessary to group items in File Explorer by 
the custom sort plugin. So far so good.
The potential challenge could be how the user knows the exact name of icon to enter in sortspec YAML. For example, the visually selected icon `AlarmLine` of the `(Ri)` pack is internally stored (by the `obsidian-icon-folder`) under the name `RiAlarmLine`. 

thanks for your suggestion!
It looks like it could be turned into a new feature of the custom sort plugin

BTW I'll become unavailable for about two weeks from now, thus not responding in this thread

**Comment by a-rock-lobster** [comment]
Super, thanks for following up so quickly!

I think that if the icon names are atleast consistant that shouldn't be a problem i.e. if the icon pack name, such as Ri, is always at the front as in your example.

**Comment by SebastianMC** [comment]
Hi @a-rock-lobster,

I did a deep dive research of how the `obsidian-icon-folder` works and I found it troublesome to determine if an item display an icon in File Explorer. This is because the icon can be displayed as a result of any of the below conditions:
- (A) icon configured directly on the folder or file -> handling is easy peasy lemon squeezy 
- (B) icon configured on a parent folder as 'inherit icon' -> handling is easy peasy as well
- (C) icon configured on some (any) of ancestor folders as 'inherit icon' -> mhm... things are getting not so trivial
- (D) icon configured in plugin settings '_Custom Icon Rules_' via a regexp or string pattern with an additional designation to match folders, files or both -> here begins the horror

For me it looks that the integration with `obsidian-icon-folder` plugin should be done in a fully-fledged fashion, that is handling of all (A), (B), (C) and (D) cases. Partial integration (e.g. ignoring the (D) case) would result in a behavior confusing to the user: s/he sees an icon on an item in Folder Explorer and at the same time the item is ignored by a seemingly correct custom sorting-spec.

In turn, knowing that the custom sorting plugin should correctly recognize (A), (B), (C) and (D) cases means that the logic of `obsidian-icon-folder` plugin should be reproduced, especially for (D). This is definitely not the way to go.

I will inquiry with the author of `obsidian-icon-folder` plugin (@FlorianWoelki) if he can expose some simple integration API on his plugin to be consumed by other plugins (like custom sorting). Seeing the long list of [open tickets](https://github.com/FlorianWoelki/obsidian-icon-folder/issues) of 'obsidian-icon-folder' doesn't look promising.

Anyways, the next step is to open a new ticket there asking for exposing of the new API for integration.

**Comment by FlorianWoelki** [comment]
> Hi @a-rock-lobster,
> 
> 
> 
> I did a deep dive research of how the `obsidian-icon-folder` works and I found it troublesome to determine if an item display an icon in File Explorer. This is because the icon can be displayed as a result of any of the below conditions:
> 
> - (A) icon configured directly on the folder or file -> handling is easy peasy lemon squeezy 
> 
> - (B) icon configured on a parent folder as 'inherit icon' -> handling is easy peasy as well
> 
> - (C) icon configured on some (any) of ancestor folders as 'inherit icon' -> mhm... things are getting not so trivial
> 
> - (D) icon configured in plugin settings '_Custom Icon Rules_' via a regexp or string pattern with an additional designation to match folders, files or both -> here begins the horror
> 
> 
> 
> For me it looks that the integration with `obsidian-icon-folder` plugin should be done in a fully-fledged fashion, that is handling of all (A), (B), (C) and (D) cases. Partial integration (e.g. ignoring the (D) case) would result in a behavior confusing to the user: s/he sees an icon on an item in Folder Explorer and at the same time the item is ignored by a seemingly correct custom sorting-spec.
> 
> 
> 
> In turn, knowing that the custom sorting plugin should correctly recognize (A), (B), (C) and (D) cases means that the logic of `obsidian-icon-folder` plugin should be reproduced, especially for (D). This is definitely not the way to go.
> 
> 
> 
> I will inquiry with the author of `obsidian-icon-folder` plugin (@FlorianWoelki) if he can expose some simple integration API on his plugin to be consumed by other plugins (like custom sorting). Seeing the long list of [open tickets](https://github.com/FlorianWoelki/obsidian-icon-folder/issues) of 'obsidian-icon-folder' doesn't look promising.
> 
> 
> 
> Anyways, the next step is to open a new ticket there asking for exposing of the new API for integration.
> 
> 

Hey,

I have currently other priorities, that's why the amount of open issues are quite huge :D
However, I was also thinking about that and creating an issue in my repository would make sense.

**Comment by SebastianMC** [comment]
Hi @FlorianWoelki 

here you go with the technical feature request:
[#155: [Technical feature request] Expose synchronous API function - getting icon name by path of folder or file](https://github.com/FlorianWoelki/obsidian-icon-folder/issues/155)

At the same time, to proceed with the integration I will employ a simple approach here in **File Explorer Custom Sorting plugin**: read the basic data of the **obsidian-icon-folder** plugin - icon names configured directly as strings or in the [iconName field](https://github.com/FlorianWoelki/obsidian-icon-folder/blob/fd9c7df1486744450cec3d7ee9cee2b34d008e56/src/main.ts#L25) of [FolderIconObject structure](https://github.com/FlorianWoelki/obsidian-icon-folder/blob/fd9c7df1486744450cec3d7ee9cee2b34d008e56/src/main.ts#L24). This integration approach would obviously break upon any related internal change of **obsidian-icon-folder** plugin, yet for the interim it looks like a viable solution

**Comment by SebastianMC** [comment]
Hi @a-rock-lobster 

the integration is completed and available in the [1.7.0 release of the plugin](https://github.com/SebastianMC/obsidian-custom-sort/releases/tag/1.7.0)

For now only the icons configured explicitly (in `obsidian-icon-folder`) on folders and files are handled, no support for icons visible in result of inheritance or regexp-rules (for details see documentation of the mentioned plugin)

The syntax of sorting spec is based on the new token `with-icon:`, with or without parameters.

Basic examples:

```yaml
---
sorting-spec: |
 target-folder: /
 with-icon:
---
```

The above example will push to the top of vault root folder all files and folders with icons configured via the `obsidian-icon-folder` plugin, directly on these folders and files. Items with icons will be sorted alphabetically (which is the default for the custom sort plugin)

```yaml
---
sorting-spec: |
 target-folder: /
 /:files with-icon:
   > modified
---
```

The above example will push to the top of the vault root folder all files with icons configured via the `obsidian-icon-folder` plugin, directly on these folders and files. Notes with icons will be sorted by modification date ('freshness')

```yaml
---
sorting-spec: |
 target-folder: Development
 with-icon: RiNumber9
 /files:folders
 with-icon: RiNumber1
   > advanced modified
 with-icon:
---
```

The above example is a bit more complex:
- it will push to the top of the _Development_ folder all files and folders with the icon `RiNumber9` configured via the `obsidian-icon-folder` plugin, directly on these folders and files.
  - items within this group will employ the default order, which is alphabetical, files and folders treated equally
  - the icon name `RiNumber9` comes from the `remix-icons` icon pack, one of packs available to install in `obsidian-icon-folder`. The icon name is visible when you use the UI of that plugin to configure the icon on a file of folder
- then will go files and folders not matching any other sorting rule of the _Development_ folder -> the `/files:folders` token w/o any parameters has that meaning
  - items within this group will employ the default order, which is alphabetical, files and folders treated equally 
  - to actually determine which items will fall in this group you need to determine all of the defined sorting rules, including the ones defined below the line with `/files:folders` 
- then will go all files and folders with the icon `RiNumber1` configured via the `obsidian-icon-folder` plugin, directly on these folders and files.
  - items within this group will employ the 'by freshness' order (`> advanced modified`), files and folders treated equally 
- and eventually, pushed to the bottom, will go files and folders with any other icon
  - for clarity: files and folder not having any icon will be captured by the `/files:folders` rule listed before

> Note, that to refresh the sorting after adding or removing icons on some folders and files you need to toggle the ribbon icon or issue the `sort-on` command in Obsidian commands palette

Please let me know if you have further questions.

BTW I'm also curious if you apply icons in your vault(s) via icon inheritance and regexp, or only via direct assignment on a file or folder.

**Comment by a-rock-lobster** [comment]
Hi @SebastianMC 

Amazing stuff. I really appreciate the effort that you have gone to. 

I mostly use direct assignment, but I have a few regexp. For example, I'm writing a monograph thesis and each top level folder (a chapter) has some common files: references, abstract, tasks, overview (mindmap) etcetera. These are regexp and I have been using sortspec to push them to the top or, in the case of the references file, the bottom using target-folder: /*. 

For now, however, I will scrub the regexp and manually set the icons so that I can have a target-folder: /* set up with common patterns of icon I use in my folder structure. Ofcourse, it would be awesome if one day sortspec worked with regexp so that reoccuring file types (i.e. references) just automatically were pushed to the top or bottom, but this implementation is already super helpful. 

Thanks again for your efforts!

**Comment by a-rock-lobster** [comment]
Hi @SebastianMC,

This is awesome! Thank you for putting in so much work. It's really appreciated 🙏

I do use some regexp but mostly it's direct assignment. I can imagine a use case when regexp working be nice but the implementation already is super helpful. 

Thanks again!

(deleted last after posting because I realised I had gotten confused - still working out the more advanced ways of using the sortspec)

**Comment by danteali** [comment]
Sorry for commenting on a closed issue but I've just installed the Icon Folder plugin as I saw that you added support.

I could be missing something completely obvious, but in the current version of Icon Folder it looks like it's no longer possible to determine the icon name in the format used by Custom Sort. When I assign an icon it displays the name of the descriptive name of the icon (e.g. `Ubuntu (Fab)`) instead of the `IconPackNumber##` which Custom Sort appears to use.

Any idea how I can find the correct icon name to be able to sort on it?

**Comment by SebastianMC** [comment]
Hi @danteali,

the name of the icon to use in sorting spec comprises the pack short name concatenated with the icon name.
For example:
- use `FabUbuntu` for icon labeled `Ubuntu (Fab)` inside the Icon Folder plugin
  - `  with-icon: FabUbuntu`  
- use `FabAndroid` for `Android (Fab)`
  - `  with-icon: FabAndroid`
and so on.

I'm aware that the documentation is still missing for this particular area :-|

This works with the most recent [(1.7.0) release of Icon Folder](https://github.com/FlorianWoelki/obsidian-icon-folder/releases/tag/1.7.0) plugin

**Comment by a-rock-lobster** [comment]
Hi, @SebastianMC 

I think that the latest update might have broken the with-icon: support? Atleast, it stopped working after I updated and I haven't been able to find any causes myself.

**Comment by SebastianMC** [comment]
Hi @a-rock-lobster,

thanks for the report, I will take a look as soon as possible

**Comment by SebastianMC** [comment]
A bug indeed, fixed in the recent [1.9.2](https://github.com/SebastianMC/obsidian-custom-sort/releases/tag/1.9.2)
Please give it a try

**Comment by a-rock-lobster** [comment]
Looks good to me :)
